### PR TITLE
Mutex/list

### PR DESCRIPTION
**Pull request by nfarah86**
@coyle this is a WIP- but was hoping to get an initial review first.

---
This change is [Reviewable](https://reviewable.io/reviews/storj/storj/254)

### CODE REVIEW

**Comment by nfarah86** [comment]
@coyle i think this is ready for review- when you get a chance. I redid the definition of list and made sure the everything was compatible with the new definition.

File: internal/test/storage.go [hunk 1]
```diff
@@ -94,29 +94,46 @@ func (m *MockKeyValueStore) Delete(key storage.Key) error {
...
-	if startingKey == nil {
+	listItems := storage.Items{}
+
+	if opts.Start == nil {
 		started = true
 	}
 	for _, key := range keySlice {
-		if !started && key == string(startingKey) {
+		if !started && key == string(opts.Start) {
 			keys = append(keys, storage.Key(key))
```
Comment (jtolds) [inline internal/test/storage.go]: can we get rid of the keys slice entirely?

File: internal/test/storage.go [hunk 2]
```diff
@@ -94,29 +94,46 @@ func (m *MockKeyValueStore) Delete(key storage.Key) error {
...
 			started = true
 			continue
 		}
 		if started {
-			if len(keys) == int(limit) {
+			if len(keys) == int(opts.Limit) {
```
Comment (jtolds) [inline internal/test/storage.go]: this function always returns false for more. shouldn't it return true in this case?

File: pkg/kademlia/routing.go [hunk 3]
```diff
@@ -112,10 +112,18 @@ func (rt *RoutingTable) GetBucket(id string) (bucket dht.Bucket, ok bool) {
...
+		Limit: 0,
+	})
+
 	if err != nil {
 		return bs, RoutingErr.New("could not get bucket ids %s", err)
 	}
+
+	kbuckets := kbucketsItems.GetKeys()
+
 	for _, v := range kbuckets {
```
Comment (jtolds) [inline pkg/kademlia/routing.go]: this for loop needs to be changed to actually use the pagination kadBucketDB.List now supports

Comment (jtolds) [reply pkg/kademlia/routing.go]:
```
listOpts := storage.ListOptions{Limit: 100}
for {
  kbucketItems, more, err := rt.kadBucketDB.List(listOpts)
  if err != nil { 
    return bs, err
  }
  for _, v := range kbucketsItems.GetKeys() {
    ...
  }
  if !more { 
    break
  }
  listOpts.Start = kbucketItems[len(kbucketItems)-1]
}
```

File: internal/test/storage.go [hunk 4]
```diff
@@ -94,29 +94,46 @@ func (m *MockKeyValueStore) Delete(key storage.Key) error {
...
 
-	if startingKey == nil {
+	listItems := storage.Items{}
+
+	if opts.Start == nil {
 		started = true
 	}
 	for _, key := range keySlice {
-		if !started && key == string(startingKey) {
+		if !started && key == string(opts.Start) {
```
Comment (jtolds) [inline internal/test/storage.go]: so, if someone provides a StartKey, it should return every key after it, not including it? it's a noninclusive start?

File: pkg/kademlia/routing.go [hunk 5]
```diff
@@ -142,10 +150,17 @@ func (rt *RoutingTable) FindNear(id dht.NodeID, limit int) ([]*proto.Node, error
 		return []*proto.Node{}, RoutingErr.New("could not get key from rt %s", err)
 	}
 	// if id is not in the routing table
-	nodeIDs, err := rt.nodeBucketDB.List(nil, 0)
+	nodeItems, _, err := rt.nodeBucketDB.List(storage.ListOptions{
+		Start: nil,
+		Limit: 0,
+	})
```
Comment (jtolds) [inline pkg/kademlia/routing.go]: same here, all of the list options should be extended to use pagination loops

File: pkg/kademlia/routing_helpers.go [hunk 6]
```diff
@@ -174,10 +175,16 @@ func (rt *RoutingTable) createOrUpdateKBucket(bucketID storage.Key, now time.Tim
 // getKBucketID: helper, returns the id of the corresponding k bucket given a node id.
 // The node doesn't have to be in the routing table at time of search
 func (rt *RoutingTable) getKBucketID(nodeID storage.Key) (storage.Key, error) {
-	kadBucketIDs, err := rt.kadBucketDB.List(nil, 0)
+	kadBucketItems, _, err := rt.kadBucketDB.List(storage.ListOptions{
+		Start: nil,
+		Limit: 0,
+	})
```
Comment (jtolds) [inline pkg/kademlia/routing_helpers.go]: needs pagination

File: pkg/kademlia/routing_helpers.go [hunk 7]
```diff
@@ -235,10 +242,20 @@ func xorTwoIds(id []byte, comparisonID []byte) []byte {
 
 // nodeIsWithinNearestK: helper, returns true if the node in question is within the nearest k from local node
 func (rt *RoutingTable) nodeIsWithinNearestK(nodeID storage.Key) (bool, error) {
-	nodes, err := rt.nodeBucketDB.List(nil, 0)
+	nodeItems, _, err := rt.nodeBucketDB.List(storage.ListOptions{
+		Start: nil,
+		Limit: 0,
+	})
```
Comment (jtolds) [inline pkg/kademlia/routing_helpers.go]: same

File: pkg/kademlia/routing_helpers.go [hunk 8]
```diff
@@ -290,10 +307,22 @@ func (rt *RoutingTable) getNodeIDsWithinKBucket(bucketID storage.Key) (storage.K
 	left := endpoints[0]
 	right := endpoints[1]
 	var nodeIDs storage.Keys
-	allNodeIDs, err := rt.nodeBucketDB.List(nil, 0)
+
+	allNodeItems, _, err := rt.nodeBucketDB.List(storage.ListOptions{
+		Start: nil,
+		Limit: 0,
+	})
```
Comment (jtolds) [inline pkg/kademlia/routing_helpers.go]: same.

either bucketDB should not have a pagination interface, or it should and we should use it.

File: pkg/overlay/server.go [hunk 9]
```diff
@@ -104,7 +104,15 @@ func (o *Server) getNodes(ctx context.Context, keys storage.Keys) ([]*proto.Node
 
 func (o *Server) populate(ctx context.Context, starting storage.Key, maxNodes, restrictedBandwidth, restrictedSpace int64) ([]*proto.Node, storage.Key, error) {
 	limit := storage.Limit(maxNodes) * 2
-	keys, err := o.cache.DB.List(starting, limit)
+	opts := storage.ListOptions{
+		Start: starting,
+		Limit: limit,
+	}
+
+	items, _, err := o.cache.DB.List(opts)
```
Comment (jtolds) [inline pkg/overlay/server.go]: same, also, you need to check the err immediately

File: pkg/pointerdb/pointerdb.go [hunk 10]
```diff
@@ -122,8 +122,15 @@ func (s *Server) List(ctx context.Context, req *pb.ListRequest) (resp *pb.ListRe
...
 
+	opts := storage.ListOptions{
+		Start: []byte(req.GetPrefix() + "/"),
+		Limit: 0,
+	}
+
 	// TODO(kaloyan): here we query the DB without limit. We must optimize it!
-	keys, err := s.DB.List([]byte(req.GetPrefix()+"/"), 0)
+	listItems, _, err := s.DB.List(opts)
+	keys := listItems.GetKeys()
```
Comment (jtolds) [inline pkg/pointerdb/pointerdb.go]: gotta check the err first, also pagination.

File: storage/boltdb/client.go [hunk 11]
```diff
@@ -95,10 +95,43 @@ func (c *Client) Get(pathKey storage.Key) (storage.Value, error) {
...
+	if err != nil {
+		return nil, false, err
+	}
+	listItems, err := c.makeListItems(paths)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return listItems, false, err
+}
```
Comment (jtolds) [inline storage/boltdb/client.go]: this never returns true for storage.More?

File: storage/common.go [hunk 12]
```diff
@@ -22,16 +22,42 @@ type Keys []Key
 // Values is the type for a slice of Values in a `KeyValueStore`
 type Values []Value
 
+// ListOptions are items that are optional for the LIST method
+type ListOptions struct {
+	Prefix       Key
+	Start        Key
```
Comment (jtolds) [inline storage/common.go]: based on the logic you had before, perhaps this should be named StartAfter?

File: storage/common.go [hunk 13]
```diff
@@ -22,16 +22,42 @@ type Keys []Key
...
+	End          Key
+	Delimeter    string
```
Comment (jtolds) [inline storage/common.go]: instead of `Delimiter string` we should have `Recursive bool` i think

File: storage/common.go [hunk 14]
```diff
@@ -22,16 +22,42 @@ type Keys []Key
 // Values is the type for a slice of Values in a `KeyValueStore`
 type Values []Value
 
+// ListOptions are items that are optional for the LIST method
+type ListOptions struct {
+	Prefix       Key
+	Start        Key
+	End          Key
```
Comment (jtolds) [inline storage/common.go]: should this be EndBefore?

File: storage/redis/client.go [hunk 15]
```diff
@@ -84,32 +84,46 @@ func (c *Client) Put(key storage.Key, value storage.Value) error {
...
+			Value: value,
+			// right now, we'll set this to false. PR2 will fix this
+			IsPrefix: false,
+		}
+
+		listItems = append(listItems, listItem)
 	}
 
-	return listKeys, nil
+	return listItems, false, nil
```
Comment (jtolds) [inline storage/redis/client.go]: storage.More should sometimes be true.

**Comment by iglesiasbrandon** [comment]
closing PR, It is no longer needed as a reference.

**Pull request closed without merging**